Fix team sign-up routing

vaqwerty2 opened the pull request:
For now, I have fixed the routing issue that emerged when the "Team Sign Up" button was clicked. It now properly renders the form, and the routing issue is resolved. Also, the 2FA's frontend implementation for now.

s-bonnor commented on home/templates/accounts/sign-in.html:
```diff
@@ -9,71 +9,86 @@
+                            <!-- End of Form -->
+
+                            <!-- 2FA Input -->
+                            <div class="form-group mb-4">
+                                <label for="2fa-code">2FA Code</label>
+                                <div class="input-group">
+                                    <span class="input-group-text" id="basic-addon2">
+                                        <span class="fas fa-key"></span>
+                                    </span>
+                                    <input type="text" id="2fa-code" name="2fa_code" class="form-control" placeholder="Enter your 2FA code">
```
the site is unable to actually send 2FA codes in a development environment so this prevents logging in
could you fix this, perhaps print the 2FA code to the terminal when DEBUG=true

s-bonnor commented on home/templates/accounts/sign-in.html:
```diff
@@ -9,71 +9,86 @@
+                            {% endfor %}
+                            <!-- End of Form -->
+
+                            <!-- 2FA Input -->
+                            <div class="form-group mb-4">
+                                <label for="2fa-code">2FA Code</label>
+                                <div class="input-group">
+                                    <span class="input-group-text" id="basic-addon2">
+                                        <span class="fas fa-key"></span>
+                                    </span>
+                                    <input type="text" id="2fa-code" name="2fa_code" class="form-control" placeholder="Enter your 2FA code">
```
the site is unable to actually send 2FA codes in a development environment so this prevents logging in
could you fix this, perhaps print the 2FA code to the terminal when DEBUG=true

vaqwerty2 replied:
Have fixed this in my next commit. Now the code is being generated and the email is being sent.

s-bonnor commented on home/forms.py:
```diff
@@ -49,16 +48,36 @@ def clean_password1(self):
         return email
     # .......................................
 
+    def set_password(self, raw_password):
+        """
+        Hashes the password using Django's password hashing system and returns it.
+        """
+        return make_password(raw_password)
+
+    def save(self, commit=True):
+        """
+        Overrides the save method to hash the password before saving the user instance.
+        """
+        user = super().save(commit=False)
+        raw_password = self.cleaned_data.get('password1')
+        user.password = self.set_password(raw_password)
+        if commit:
+            user.save()
+        return user
```
This is already handled by the UserModel, the form does not need to handle hashing the password, and in fact should not handle hashing the password

vaqwerty2 replied:
Have made the relevant change in the next commit

s-bonnor commented on core/settings.py:
```diff
@@ -136,6 +136,15 @@
 # Password validation
 # https://docs.djangoproject.com/en/4.1/ref/settings/#auth-password-validators
 
+# Password hashing using bcrypt
+PASSWORD_HASHERS = [
+    'django.contrib.auth.hashers.BCryptSHA256PasswordHasher',  # Built-in bcrypt with SHA256
+    'django.contrib.auth.hashers.Argon2PasswordHasher',
+    'django.contrib.auth.hashers.PBKDF2PasswordHasher',
+    'django.contrib.auth.hashers.PBKDF2SHA1PasswordHasher',
+]
```
can you provide some justification for using bcrypt over the django framework default of pbkdf2?
also BCryptSHA256PasswordHasher fails to load due to a missing library

vaqwerty2 replied:
The reason for using BCryp (To enhance password security): 
1. Stronger Defense Against Hardware Cracking
bcrypt has a design that is less susceptible to hardware-accelerated attacks, such as those leveraging GPUs or ASICs, because of its reliance on a memory-hard key derivation process. PBKDF2 lacks memory-hardness, making it easier to parallelize on modern hardware.
2. Built-in Salting Mechanism
bcrypt automatically includes a salt when hashing, ensuring that even identical passwords result in completely different hashes. PBKDF2 also supports salting, but it requires developers to manage it correctly, increasing the risk of implementation errors.
3. Memory Usage
bcrypt has a lower memory requirement compared to some PBKDF2 implementations, making it more efficient in resource-constrained environments.

vaqwerty2 replied:
Evidence of BCrypt working:
<img width="712" alt="Screenshot 2024-12-04 at 14 48 23" src="https://github.com/user-attachments/assets/e24e3f20-a530-45eb-be81-e293336bfe3c">

s-bonnor commented on home/views.py:
```diff
@@ -290,31 +293,62 @@ def password_gen(request):
+                        from_email="deakinhardhatwebsite@gmail.com",
+                        recipient_list=[email],
+                        fail_silently=False,
+                    )
+                    print(f"OTP sent to {email}.")
+
+                    # Redirect to verify token page with context
+                    messages.success(request, "Account created successfully! Check your email for the OTP.")
+                    return render(
+                        request,
+                        'accounts/verify_token.html',
+                        {'otp': otp, 'first_name': user.first_name, 'last_name': user.last_name, 'email': user.email}
+                    )
+                except Exception as e:
+                    print(f"Error saving user or sending email: {e}")
+                    print(traceback.format_exc())  # Print detailed traceback
+                    messages.error(request, "An error occurred while creating the account. Please try again.")
+            else:
+                print("Form is invalid. Errors:")
+                print(form.errors)  # Debugging log for form errors
+                messages.error(request, "Please fix the errors below.")
         else:
-            print("Registration failed!")
-    else:
-        form = RegistrationForm()
- 
-    context = { 'form': form }
-    return render(request, 'pages/index.html', context)
+            print("GET request received for registration.")
+            form = RegistrationForm()
+
+        context = {'form': form}
+        return render(request, 'accounts/sign-up.html', context)
+
+    except Exception as e:
+        # Catch any unexpected errors and print to the terminal
+        print(f"Unexpected error in register view: {e}")
+        print(traceback.format_exc())
+        messages.error(request, "An unexpected error occurred. Please try again later.")
+        return render(request, 'accounts/sign-up.html', {'form': RegistrationForm()})
```
currently the register view is unused and mostly a duplicate of the signup view, could you merge them into 1 view

vaqwerty2 replied:
Have merged them now

s-bonnor reviewed (changes requested):
biggest issue is currently its impossible to register an account and login, because the login form asks for a OTP code which never gets generated

vaqwerty2 commented:
Hey Sam, I have addressed all the comments and have made the relevant changes

vaqwerty2 commented:
Fixed the routing issue with the Team Sign-Up button. Collaborated with the backend team to resolve the regex issue during account creation, ensuring user details are successfully saved to the local database. Enhanced security by hashing stored passwords using bcrypt. Implemented an OTP system, sending a one-time password via email for account verification during signup. For login, credentials are verified, followed by another OTP sent via email for additional login verification. Introduced MFA after OTP verification using a CAPTCHA, which, upon successful validation, logs the user into the website. As evident by the preferences page displayed post-login, as demonstrated in this video : [](https://deakin.au.panopto.com/Panopto/Pages/Viewer.aspx?id=b5434dc3-2706-46de-b658-b23e00f6c0d4)

s-bonnor reviewed (changes requested):
Looks good, just need you to merge the current main branch in and fix any merge conflicts that happen

s-bonnor reviewed (approved):
All looks good now
